arch/arm/core/aarch64/reset.S

```diff
@@ -19,131 +13,93 @@
 _ASM_FILE_PROLOGUE
 
 /*
- * Platform may do platform specific init at EL3.
- * The function implementation must preserve callee saved registers as per
- * AArch64 ABI PCS.
+ * Set the minimum necessary to safely call C code
  */
 
-WTEXT(z_arch_el3_plat_init)
-SECTION_FUNC(TEXT,z_arch_el3_plat_init)
-ret
+GTEXT(__reset_prep_c)
+SECTION_SUBSEC_FUNC(TEXT,_reset_section,__reset_prep_c)
+	switch_el x0, 3f, 2f, 1f
+3:
+	/* Reinitialize SCTLR from scratch in EL3 */
+	ldr	w0, =(SCTLR_EL3_RES1 | SCTLR_I_BIT | SCTLR_SA_BIT)
+	msr	sctlr_el3, x0
+	b	out
+2:
+	/* Disable alignment fault checking */
+	mrs	x0, sctlr_el2
+	bic	x0, x0, SCTLR_A_BIT
+	msr	sctlr_el2, x0
+	b	out
+1:
+	/* Disable alignment fault checking */
+	mrs	x0, sctlr_el1
+	bic	x0, x0, SCTLR_A_BIT
+	msr	sctlr_el1, x0
+
+out:
+	isb
+
+	/* Select SP_EL0 */
+	msr	SPSel, #0
+
+	/* Initialize stack */
+	ldr	x0, =(z_interrupt_stacks)
+	add	x0, x0, #(CONFIG_ISR_STACK_SIZE)
+	mov	sp, x0
+
+	ret
```

Comment:
I like how this file came out.

Looking at my current local changes, I think I'd still have to override `__start`, unfortunately, as we have to do a little bit (~3 lines) of platform-specific setup before entering C-land.

Can we split out the EL3/2/1 parts of `__reset_prep_c` into overridable asm functions (or alternatively add overridable nop-calls here)? (This would require that `__reset_prep_c` save/restore its return value (likely into a different register, as the stack isn't yet set up), or tail-call into the overridable function.)

Reply:
Yup, no prob.